### PR TITLE
Fixed #27515 -- Make AuthenticationForm use the username length from the user model

### PR DESCRIPTION
**Pull request by RaminFP**
[ticket](https://code.djangoproject.com/ticket/27515)

### CODE REVIEW

File: django/contrib/auth/forms.py [hunk 1]
```diff
@@ -178,9 +178,12 @@ def __init__(self, request=None, *args, **kwargs):
         self.user_cache = None
         super(AuthenticationForm, self).__init__(*args, **kwargs)
 
-        # Set the label for the "username" field.
+        # Set the label and username length for the "username" field.
         UserModel = get_user_model()
         self.username_field = UserModel._meta.get_field(UserModel.USERNAME_FIELD)
+        self.fields['username'].widget.attrs['maxlength'] = UserModel._meta.get_field(
```
Comment (charettes) [inline django/contrib/auth/forms.py]: Use a local variable to store `UserModel._meta.get_field(UserModel.USERNAME_FIELD)` as it's used twice.

File: django/contrib/auth/forms.py [hunk 2]
```diff
@@ -178,9 +178,12 @@ def __init__(self, request=None, *args, **kwargs):
...
+            UserModel.USERNAME_FIELD
+        ).max_length
```
Comment (charettes) [inline django/contrib/auth/forms.py]: `max_length` can be `None` for non-`CharField` instances of `Field`. You might want to assign the attribute only if it is not `None`.

Comment (RaminFP) [inline django/contrib/auth/forms.py]: @charettes Thanks for review,

File: django/contrib/auth/forms.py [hunk 3]
```diff
@@ -178,9 +178,11 @@ def __init__(self, request=None, *args, **kwargs):
         self.user_cache = None
         super(AuthenticationForm, self).__init__(*args, **kwargs)
 
-        # Set the label for the "username" field.
+        # Set the label and username length for the "username" field.
```
Comment (RaminFP) [inline django/contrib/auth/forms.py]: `username length` to `max_length`

**Comment by timgraham** [comment]
This doesn't look correct since modifying only the widget value won't enforce server-side validation. Also tests and documentation updates are required.

**Comment by RaminFP** [comment]
@timgraham are you sure validate and docs? https://github.com/django/django/commit/780bddf75b93784470a2e352ed44ee35a751d667

**Comment by timgraham** [comment]
Yes please.

**Comment by RaminFP** [comment]
@timgraham Thank you so much,still i not sure for docs what is needed, please test and feedback, 
Thanks again,

File: django/contrib/auth/forms.py [hunk 4]
```diff
@@ -189,17 +192,26 @@ def clean(self):
         password = self.cleaned_data.get('password')
 
         if username is not None and password:
-            self.user_cache = authenticate(self.request, username=username, password=password)
-            if self.user_cache is None:
-                raise forms.ValidationError(
-                    self.error_messages['invalid_login'],
-                    code='invalid_login',
-                    params={'username': self.username_field.verbose_name},
+            if len(username) <= self.username_field.max_length:
```
Comment (timgraham) [inline django/contrib/auth/forms.py]: It looks incorrect to reimplement a field's validation in `clean()`. Why isn't the field validating the max_length?

Comment (RaminFP) [inline django/contrib/auth/forms.py]: @timgraham i going to avoid before call `authenticate` for check,and return error length message,

Comment (RaminFP) [inline django/contrib/auth/forms.py]: your means i use this : https://docs.djangoproject.com/en/1.10/_modules/django/core/validators/#MaxLengthValidator

Comment (timgraham) [inline django/contrib/auth/forms.py]: No, I mean that `clean_<fieldname>()` should take care of it. https://docs.djangoproject.com/en/stable/ref/forms/validation/

**Comment by RaminFP** [comment]
@timgraham Thanks of your patience, please review

File: django/contrib/auth/forms.py [hunk 5]
```diff
@@ -178,16 +178,26 @@ def __init__(self, request=None, *args, **kwargs):
...
         self.username_field = UserModel._meta.get_field(UserModel.USERNAME_FIELD)
+        if self.username_field.max_length is not None:
+            self.fields['username'].max_length = self.username_field.max_length
+            self.fields['username'].widget.attrs['maxlength'] = self.username_field.max_length
         if self.fields['username'].label is None:
             self.fields['username'].label = capfirst(self.username_field.verbose_name)
 
+    def clean_username(self):
+        username = self.cleaned_data.get('username')
+        if len(username) > self.username_field.max_length:
```
Comment (RaminFP) [inline django/contrib/auth/forms.py]: This is normal condition on clean method,for check greater length username  of 150 characters,

**Comment by RaminFP** [comment]
@timgraham change code ,
display error,if you think issue on style code , please modify PR, Thanks
```
Please correct the error below.

Ensure this value has at most 150 characters (it has 200).

```

File: django/contrib/auth/forms.py (hunk 5 above)
Comment (RaminFP) [inline django/contrib/auth/forms.py]: you know this messge `{'f': ['Ensure this value has at most 1 character (it has 2).']}` isn't crystal for user which fields(username or password) error

File: django/contrib/auth/forms.py [hunk 6]
```diff
@@ -160,7 +160,9 @@ class AuthenticationForm(forms.Form):
         strip=False,
         widget=forms.PasswordInput,
     )
-
+    username = forms.CharField(
```
Comment (timgraham) [inline django/contrib/auth/forms.py]: This is, we already have `username = UsernameField`.

File: django/contrib/auth/forms.py [hunk 7]
```diff
@@ -178,16 +180,18 @@ def __init__(self, request=None, *args, **kwargs):
         self.user_cache = None
         super(AuthenticationForm, self).__init__(*args, **kwargs)
 
-        # Set the label for the "username" field.
+        # Set the label and maxlength for the "username" field.
         UserModel = get_user_model()
         self.username_field = UserModel._meta.get_field(UserModel.USERNAME_FIELD)
+        if self.username_field.max_length is not None:
```
Comment (timgraham) [inline django/contrib/auth/forms.py]: How could `username_field.max_length` be `None`?

File: django/contrib/auth/forms.py [hunk 8]
```diff
@@ -178,16 +180,18 @@ def __init__(self, request=None, *args, **kwargs):
...
+            self.fields['username'].max_length = self.username_field.max_length
+            self.fields['username'].widget.attrs['maxlength'] = self.username_field.max_length
         if self.fields['username'].label is None:
             self.fields['username'].label = capfirst(self.username_field.verbose_name)
 
     def clean(self):
         username = self.cleaned_data.get('username')
         password = self.cleaned_data.get('password')
-
```
Comment (timgraham) [inline django/contrib/auth/forms.py]: Please don't make unrelated whitespace changes.

File: docs/topics/auth/default.txt [hunk 9]
```diff
@@ -1616,7 +1616,7 @@ provides several built-in forms located in :mod:`django.contrib.auth.forms`:
 
 .. class:: AuthenticationForm
 
-    A form for logging a user in.
+    The form read the user submitted data, validate the form and take the final step of logging the user in.
```
Comment (timgraham) [inline docs/topics/auth/default.txt]: This change isn't describing anything that changed as part of this pull request.

File: tests/auth_tests/test_forms.py [hunk 10]
```diff
@@ -281,6 +281,21 @@ def test_inactive_user(self):
...
+        # The user submits an invalid maxlength ex:(200 length) username.
+        invalid_maxlength_username = "A" * 200
+        data = {
+            'username': invalid_maxlength_username,
+            'password': '123456',
+        }
+        form = AuthenticationForm(None, data)
+        self.assertFalse(form.is_valid())
+        self.assertEqual(
+            form.non_field_errors(), [
```
Comment (timgraham) [inline tests/auth_tests/test_forms.py]: This test doesn't pass. Did you try running it?

**Comment by RaminFP** [comment]
>You need to review your work better or increase your Django knowledge before contributing.
@ti bad comment,ok any contributing,
